docs/design.md

```diff
@@ -0,0 +1,116 @@
+Design
+======
+
+The main goals of libnetwork are highlighted in the [roadmap](../ROADMAP.md).
+This document describes how libnetwork has been designed in order to acheive this.
+Requirements for individual releases can be found on the [Project Page](https://github.com/docker/libnetwork/wiki)
+
+## Legacy Docker Networking
+
+Prior to libnetwork a container's networking was handled in both Docker Engine and libcontainer.
+Docker Engine was responsible for providing the configuration of the container's networking stack.
+Libcontainer would then use this information to create the necessary networking devices and move them in to a network namespace.
+This namespace would then be used when the container is started.
+
+## The Container Network Model
+
+Libnetwork implements Container Network Model (CNM) which formalizes the steps required to provide networking for containers while providing an abstraction that can be used to support multiple network drivers. The CNM is built on 3 main components.
+
+**Sandbox**
+
+A Sandbox contains the configuration of a container's network stack.
+This includes management of the container's interfaces, routing table and DNS settings. 
+An implementation of a Sandbox could be a Linux Network Namespace, a FreeBSD Jail or other similar concept.
+A Sandbox may contain *many* endpoints from *multiple* networks
+
+**Endpoint**
+
+An Endpoint joins a Sandbox to a Network.
+An implementation of an Endpoint could be a `veth` pair, an Open vSwitch internal port or similar.
+An Endpoint can belong to *only one* network but may only belong to *one* Sandbox
+
+**Network**
+
+A Network is a group of Endpoints that are able to communicate with each-other directly.
+An implementation of a Network could be a Linux bridge, a VLAN etc...
+Networks consist of *many* endpoints
+
+## API
+
+Consumers of the CNM, like Docker for example, interact through the following APIs
+
+The `NetworkController` object is created to manage the allocation of Networks and the binding of these Networks to a specific Driver
+Once a Network is created, `network.CreateEndpoint` can be called to create a new Endpoint in a given network.
+When an Endpoint exists, it can be joined to a Sandbox using `endpoint.Join(id)`. If no Sandbox exists, one will be created, but if the Sandbox already exists, the endpoint will be added there.
```

Comment:
@mrjana 
> This cannot be the case because all the SandboxInfo needs to be provided by the driver during endpoint creation time. 
> ...
> With this design endpoint is no more a local resource but a universal one across the cluster. While Join is purely a local event indicating to the driver that a container has joined the endpoint in this host.

Thanks for clarifying. I think I get the intent but I am still trying to understand the rules/assumptions under which a driver is expected to behave/implemented in this model. BTW it seems easy to mix up resource allocation which can happen at global or local level (like IP allocation can be global or local while netdevice creation think ovs internal ports will be local), so it might be helpful to clarify them here or in a section that describes the driverapi interface. For now let me try clarify a few basic things by considering a multi-host environment. For the following I consider IP address as a global resource while netdevice as a local one.

Does driver.CreateEndpoint() get called for **each** driver instance (per host)? From your description looks like not since Endpoint is not a local resource. Is it correct? Or do you expect **all** driver instances to get this trigger but then they need to figure out how to deal with it (by having their global view of state)? I think latter might be better as driver might need to have a global view anyway. In either case, does libnetwork expect the returned sandboxinfo to be filled up with Interfaces (netdevice name and IP address) and Gateway addresses etc? If so, I am a bit concerned with tying of local resources like net-device names and global-resources like IP addresses through a sandbox created for a global event, as it gives an impression that local resource (like netdevice) should have been created/reserved here as well. Also it's not clear how would this work in case of multiple drivers (IPAM driver with a vlan-based driver that  I asked about earlier).

It seems driver.Join() will be called **only** for the driver instance where the container is actually going to be started, so it is more suitable place for doing local plumbing like actual netdevice creation. Is it correct? But if sandboxinfo was filled up with interfaces and IP addresses before then why even trigger the driver here again?

Regarding the argument for services, I am wondering if they really need to be treated specially. Why can't they be treated just like any other container and endpoints joining/leaving them? If possible, can you please share a concrete example that came up during your discussion, that might help me understand.

Overall I feel at early stages, it might be desirable to keep driverapi surface to the minimum by just giving drivers either Create/Delete or Join/Leave triggers at a deterministic time in a container's lifecycle. It keeps things simple with less interpretations to worry about :blush: . It also leaves room for libnetwork to build more complex usecases (multi-host deployment; chaining of drivers etc) on top of simple, deterministic, composable network drivers. But this might be just my opinion.